pallets/capacity/src/lib.rs

```diff
@@ -0,0 +1,658 @@
+
+			Self::deposit_event(Event::EpochLengthUpdated { blocks: length });
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Checks to see if staker has sufficient free-balance to stake the minimum required staking amount,
+	/// and leave the minimum required free balance after staking.
+	///
+	/// # Errors
+	/// * [`Error::ZeroAmountNotAllowed`]
+	/// * [`Error::InvalidTarget`]
+	/// * [`Error::BalanceTooLowtoStake`]
+	///
+	fn ensure_can_stake(
+		staker: &T::AccountId,
+		target: MessageSourceId,
+		amount: BalanceOf<T>,
+	) -> Result<(StakingAccountDetails<T>, BalanceOf<T>), DispatchError> {
+		ensure!(amount > Zero::zero(), Error::<T>::ZeroAmountNotAllowed);
+		ensure!(T::TargetValidator::validate(target), Error::<T>::InvalidTarget);
+
+		let staking_account = Self::get_staking_account_for(&staker).unwrap_or_default();
+		let stakable_amount = staking_account.get_stakable_amount_for(&staker, amount);
+
+		ensure!(stakable_amount > Zero::zero(), Error::<T>::BalanceTooLowtoStake);
+
+		let new_active_staking_amount = staking_account
+			.active
+			.checked_add(&stakable_amount)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		ensure!(
+			new_active_staking_amount >= T::MinimumStakingAmount::get(),
+			Error::<T>::InsufficientStakingAmount
+		);
+
+		Ok((staking_account, stakable_amount))
+	}
+
+	/// Increase a staking account and target account balances by amount.
+	/// Additionally, it issues Capacity to the MSA target.
+	fn increase_stake_and_issue_capacity(
+		staker: &T::AccountId,
+		staking_account: &mut StakingAccountDetails<T>,
+		target: MessageSourceId,
+		amount: BalanceOf<T>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		staking_account.deposit(amount).ok_or(ArithmeticError::Overflow)?;
+
+		let capacity = Self::calculate_capacity(amount);
+		let mut target_details = Self::get_target_for(&staker, &target).unwrap_or_default();
+		target_details.deposit(amount, capacity).ok_or(ArithmeticError::Overflow)?;
+
+		let mut capacity_details = Self::get_capacity_for(target).unwrap_or_default();
+		capacity_details.deposit(&amount).ok_or(ArithmeticError::Overflow)?;
+
+		Self::set_staking_account(&staker, staking_account);
+		Self::set_target_details_for(&staker, target, target_details);
+		Self::set_capacity_for(target, capacity_details);
+
+		Ok(capacity)
+	}
+
+	/// Sets staking account details.
+	fn set_staking_account(staker: &T::AccountId, staking_account: &StakingAccountDetails<T>) {
+		T::Currency::set_lock(STAKING_ID, &staker, staking_account.total, WithdrawReasons::all());
+		StakingAccountLedger::<T>::insert(staker, staking_account);
+	}
+
+	/// Deletes staking account details
+	fn delete_staking_account(staker: &T::AccountId) {
+		T::Currency::remove_lock(STAKING_ID, &staker);
+		StakingAccountLedger::<T>::remove(&staker);
+	}
+
+	/// If the staking account total is zero we reap storage, otherwise set the acount to the new details.
+	fn update_or_delete_staking_account(
+		staker: &<T>::AccountId,
```

Comment:
Should this be `&T::AccountId` or `&<T>::AccountId`?

Reply:
https://github.com/LibertyDSNP/frequency/pull/1372/files